powerauth-java-server/src/main/java/io/getlime/security/powerauth/app/server/service/behavior/tasks/TemporaryKeyBehavior.java

```diff
@@ -0,0 +1,345 @@
+                    throw new IllegalArgumentException("Activation ID not found.");
+                }
+                final ActivationRecordEntity activation = activationWithoutLock.get();
+                if (activation.getActivationStatus() != ActivationStatus.ACTIVE || !appId.equals(activation.getApplication().getId())) {
+                    throw new IllegalArgumentException("Activation ID not found.");
+                }
+
+                final EncryptionMode encryptionMode = activation.getServerPrivateKeyEncryption();
+                final String serverPrivateKeyBase64 = activation.getServerPrivateKeyBase64();
+                final ServerPrivateKey serverPrivateKeyEncrypted = new ServerPrivateKey(encryptionMode, serverPrivateKeyBase64);
+                String decryptedServerPrivateKey = serverPrivateKeyConverter.fromDBValue(serverPrivateKeyEncrypted, activation.getUserId(), activation.getActivationId());
+                final byte[] serverPrivateKeyBytes = Base64.getDecoder().decode(decryptedServerPrivateKey);
+                final PrivateKey serverPrivateKey = keyConvertor.convertBytesToPrivateKey(serverPrivateKeyBytes);
+
+                final byte[] serverPublicKeyBytes = Base64.getDecoder().decode(activation.getDevicePublicKeyBase64());
+                final PublicKey serverPublicKey = keyConvertor.convertBytesToPublicKey(serverPublicKeyBytes);
+
+                final byte[] devicePublicKeyBytes = Base64.getDecoder().decode(activation.getDevicePublicKeyBase64());
+                final PublicKey devicePublicKey = keyConvertor.convertBytesToPublicKey(devicePublicKeyBytes);
+                final SecretKey transportKey = keyFactory.deriveTransportKey(serverPrivateKey, devicePublicKey);
+                final byte[] transportKeyBytes = keyConvertor.convertSharedSecretKeyToBytes(transportKey);
+
+                final TemporaryKeyResult result = new TemporaryKeyResult();
+                result.setSecretKeyBytes(transportKeyBytes);
+                result.setPrivateKey(serverPrivateKey);
+                result.setPublicKey(serverPublicKey);
+                return result;
+            }
+         } else {
+            throw new IllegalArgumentException("App key must be specified.");
+         }
+    }
+
+    private TemporaryPublicKeyResponseClaims generateAndStoreNewKey(TemporaryPublicKeyRequestClaims requestClaims) throws CryptoProviderException, InvalidKeyException {
+
+        // Generate a temporary key pair
+        final KeyPair temporaryKeyPair = keyGenerator.generateKeyPair();
+
+        // Prepare and store the entity
+        final TemporaryKeyEntity temporaryKeyEntity = new TemporaryKeyEntity();
+        temporaryKeyEntity.setId(UUID.randomUUID().toString());
+        temporaryKeyEntity.setAppKey(requestClaims.getAppKey());
+        temporaryKeyEntity.setActivationId(requestClaims.getActivationId());
+        temporaryKeyEntity.setPrivateKeyEncryption(EncryptionMode.NO_ENCRYPTION);
```

Comment:
This seems incorrect, the private key is always unencrypted. Shouldn't we encrypt the private key, in case encryption is enabled?